monkeytype/typing.py

```diff
@@ -26,13 +27,42 @@
 
 
 DUMMY_TYPED_DICT_NAME = 'DUMMY_NAME'
+REQUIRED_FIELD_PREFIX = 'REQUIRED__'
+NON_REQUIRED_FIELD_PREFIX = 'NON_REQUIRED__'
 
 
 # Functions like shrink_types and get_type construct new types at runtime.
 # Mypy cannot currently type these functions, so the type signatures for this
 # file live in typing.pyi.
 
 
+def make_typed_dict(*, required_fields=None, non_required_fields=None) -> type:
+    required_fields = required_fields or {}
+    non_required_fields = non_required_fields or {}
+    assert all(not key.startswith(REQUIRED_FIELD_PREFIX)
+               and not key.startswith(NON_REQUIRED_FIELD_PREFIX)
+               for key in chain(required_fields.keys(), non_required_fields.keys()))
+
+    required_fields = {(REQUIRED_FIELD_PREFIX + key): value
+                       for key, value in required_fields.items()}
+    non_required_fields = {(NON_REQUIRED_FIELD_PREFIX + key): value
+                           for key, value in non_required_fields.items()}
+    return TypedDict(DUMMY_TYPED_DICT_NAME, {
```

Comment:
I wonder if it would simplify this diff if we used our own MonkeyType-internal representation for `TypedDict` types, instead of using the actual `typing.TypedDict`. If we used an internal representation we could track required vs optional fields in a more explicit way (e.g. two separate dictionaries) instead of requiring all this string prefix stuff.

The main argument for using the actual `TypedDict` type is that it's more intuitive for people writing rewriters? But I think the presence of the key prefixes destroys that intuitiveness anyway; we might as well just use our own representation and document it.

Reply:
Yeah, my initial version was using `TypedDict({"required_fields": TypedDict(...), "non_required_fields": TypedDict(...)})`.

The main thing I wanted was to make the representation be a subtype of `type` so that it would be used seamlessly with the existing code. For example, `get_dict_type` would be able to return this `TypedDict` and we would be able to have `get_type([{"name": "foo"}])` be `List[TypedDict(...)]`, etc. Similarly, `shrink_typed_dict_types` would be able to return this `TypedDict` and we would be able to have recursive `TypedDicts`.

So, I've gone back to a `TypedDict` with one nested `TypedDict` each for required and optional fields. Let me know if that works for you. I tried defining a `class InternalTypedDict(type): ...` but I had trouble getting that to work and gave up.

Reply:
Sounds good!